### ISSUE TITLE
Add MixedRadix to replace CompoundUnit aka 'MixedUnit and similar

### ISSUE THREAD

**Issue by andi-huber** [unitsofmeasurement/indriya#211]
Thanks, that's highly appreciated.

See https://github.com/unitsofmeasurement/unit-api/issues/190

**Comment by andi-huber** [comment]
Quick thoughts on that:
1) MixedRadix also needs a method for parsing eg. '1 ft 2 in' into a `Quantity`
```java
MixedRadix<Length> mixedRadix = MixedRadix.ofPrimary(USCustomary.FOOT).mix(USCustomary.INCH);
Quantity<Length> quantity = mixedRadix.parse("1 ft 2 in");
```
2) Once we know how to format and parse with `MixedRadix`, we should be able to provide a `MixedRadixFormat` that implements `QuantityFormat`.

3) Regarding numerical precision and calculation performance, a special use-case comes to mind. Lets say we want to do time calculations (eg. only additions) with required precision of seconds. Given a time-quantity one might state that for this using integer arithmetic for the seconds should be sufficient.

Question: can we do this with integer arithmetic?

```java
MixedRadix<Time> mixedRadix = MixedRadix.ofPrimary(HOUR).mix(MINUTE).mix(SECOND);
Quantity<Time> quantity = mixedRadix.create(9, 20, 0);
Quantity<Time> increment = Quantities.getQuantity(30, SECOND);

// how to do this with only using integer arithmetic?
// problem is, that 'quantity' as created by 'mixedRadix' has Unit = HOUR
Quantity<Time> quantity.add(increment);
```

**Comment by andi-huber** [comment]
New findings after the latest refinement:

ad 3) I've introduces a new factory method `MixedRadix.ofLeastSignificantAsPrimary(Unit)`, which allows for `SECOND` to be our 'primary' unit. We then might want this test to succeed:

```java
void leastSignificantShouldDriveArithmetic() {
        
        // given
    
        MixedRadix<Time> mixedRadix = MixedRadix.ofLeastSignificantAsPrimary(HOUR).mix(MINUTE).mix(SECOND);
        
        Quantity<Time> startTime = mixedRadix.createQuantity(9, 20, 0);
        Quantity<Time> duration = Quantities.getQuantity(30, SECOND);

        // when
        
        Quantity<Time> endTime = startTime.add(duration);
        Number[] timeParts = mixedRadix.extractValues(endTime);
        
        // then
        
        assertEquals(9, timeParts[0]); // should be actually an int
        assertEquals(20, timeParts[1]); // should be actually an int
        assertEquals(30, timeParts[2]); // should be actually an int
        
    }
```
4) There is a possible surprise in creating a negative Quantiy `- (1 ft 2 in)` using the 'MixedRadix':

To do this correctly one has to write (which I don't like) ...

```java
MixedRadix<Length> mixedRadix = MixedRadix.ofPrimary(USCustomary.FOOT).mix(USCustomary.INCH);
Quantity<Length> lengthQuantity = mixedRadix.createQuantity(-1, -2);
```
Which raises the next question: How to format negative Quantities with mixed-radix? Where to put the sign? Use parenthesis?

**Comment by keilw** [comment]
@andi-huber Thanks for the PR so far. Would it be possible to revisit the ideas from https://github.com/unitsofmeasurement/unit-api/issues/185 keeping `MixedRadix` and `MixedRadixFormat` separate?

Instead of harder to understand ofPrimary() or ofLeastSignificantAsPrimary() factories, how about something like:
```java
MixedRadix<Length> mixedRadix = MixedRadix.of(USCustomary.FOOT).mix(USCustomary.INCH).setPrimary(USCustomary.INCH);
``` 
`of()` might also take both a single `Unit` and as an alternative a vargarg of units. Looking at Java 9 collections like `List` have 11 `of()` methods with different variations of the same argument type from 1 to variable numbers. @otaviojava mentioned it's an effort to improve the performance when creating a list, but it looks utterly confusing to me from a design point. Nevertheless, if we give `MixedRadix` two or three `of()` methods it won't be as bad.

**Comment by andi-huber** [comment]
Werner, thx for your input, much appreciated! I'll see that I can further refine my current PR.

To discuss your comments ...

Suggested terminology for 'mixed-radix' on eg. HOUR:MINUTE:SECOND:
- Primary Unit ... the Unit of the Quantity we want `Quantity<Q> MixedRadix<Q>.createQuantity(...)` to return
- Leading Unit ... HOUR (most significant)
- Trailing Unit ... SECOND (least significant)

`MixedRadix.getPrimaryUnit()` returns the immutable `Primary Unit`.

So we need to design a fluent way of constructing a `MixedRadix`, that is guaranteed to have a non-null `primaryUnit` field initialized. There are multiple ways to do this (eg. by convention, or if all other attempts fail, use a Builder). Difficulty is to make this as little boilerplate as possible, while keeping the concept easy to understand. Apart from the ugly naming, at least, that's conceptually the best I could come up with ...

- `MixedRadix.ofPrimary(Unit primaryUnit)` makes the 'leading' Unit the 'primary' Unit.
- `MixedRadix.ofLeastSignificantAsPrimary(Unit primaryUnit)` makes the 'trailing' Unit the 'primary' Unit, no matter how many mixed-radix Units we append eg. `MixedRadix.ofLeastSignificantAsPrimary(HOUR).mix(MINUTE).mix(SECOND)` makes `SECOND` the 'primary' Unit.

Can we save this concept by finding better method names?

I do like the `Builder` concept like in `NumberDelimiterQuantityFormat.Builder`, however I do have a slight preference to using a different naming scheme for the 'withers' to be distinguished from bean-convention setters. ('withers' and 'setters' should not share same method signature.) In my personal opinion 'lombok' provides the golden standard on how builders [1] should look like. What do you think?
Regardless of my preferences, I agree we should be consistent with `Builder`s, where ever these occur. 

I'm not sure whether I understand your comment (also in the PR) regarding the `QuantityFormat`. I believe the new `MixedRadixFormat` is implemented very much in the way you are suggesting. I also agree, `MixedRadix.format(...)` and `MixedRadix.parse(...)` should be refactored/moved to `MixedRadixFormat`.

One comment on `Unit<Q> ... units` as `vararg`: When we do this, we no longer have compile-time type-safety! If we need to pass a variable size of Units that have a generic-type constraint, we should resort to `Collection<Unit<Q>>` instead.

[1] https://projectlombok.org/features/Builder

**Comment by andi-huber** [comment]
Maybe this could work:

```java

// primary = defined by convention, either LEADING (=HOUR) or TRAILING (=SECOND)
// (could be configured)
MixedRadix<Time> mixedRadix = MixedRadix.of(HOUR).mix(MINUTE).mix(SECOND);

// primary = HOUR
MixedRadix<Time> mixedRadix = MixedRadix.ofPrimary(HOUR).mix(MINUTE).mix(SECOND);
        
// primary = MINUTE
MixedRadix<Time> mixedRadix = MixedRadix.of(HOUR).mixPrimary(MINUTE).mix(SECOND);

// primary = SECOND
MixedRadix<Time> mixedRadix = MixedRadix.of(HOUR).mix(MINUTE).mixPrimary(SECOND);

// IllegalStateException: a primary unit was already defined
MixedRadix<Time> mixedRadix = MixedRadix.ofPrimary(HOUR).mixPrimary(MINUTE);
```

**Comment by keilw** [comment]
Although it would require an extra step of `setPrimary()`, I believe that's more intuitive and less confusing than a single list of quantities making up the model.

https://github.com/unitsofmeasurement/indriya/blob/master/src/main/java/tech/units/indriya/quantity/CompositeQuantity.java has a `Map`, preserving the order, I see a possible chain to combine `ofPrimary(HOUR).mixPrimary(SECOND)` extremely problematic. Would it replace the first primary choice???

**Comment by andi-huber** [comment]
I see all the 5 cases covered above. Am I missing something?

**Comment by keilw** [comment]
I think we're getting closer to a workable approach. A few methods are still boilerplate, e.g. `getUnitCount()` feels a bit redundant if we guarantee, that `getUnits()` must never be null, then its `size() `method can be called directly.

**Comment by andi-huber** [comment]
- I agree, 'getUnitCount()' is just for internal convenience. 
- I have no strong opinion into which package to put `MixedRadix`, its a factory for `Quantity` but also kind of a factory for `MixedQuantityFormat`, hard to say.

In the meantime, I'm working on arithmetic optimizations.

**Comment by keilw** [comment]
About the "factory" for `MixedQuantityFormat`, that merely seems a wrapper around `MixedRadixFormat.of(this, options);` If `of()` only takes 2 arguments it could work, but IMO we shall harmonize how `MixedQantityFormat` and `NumberDelimiterQuantityFormat` are constructed. Thus a MixedQuantityFormat.Builder makes more sense based on what I did with great help from @thodorisbais in this area.

**Comment by andi-huber** [comment]
Only difference is, that this proposed `Builder` would also need to receive an instance of `MixedRadix`, for which the Builder cannot provide a default value.

**Comment by keilw** [comment]
Yes the idea would be to also pass the `MixedRadix` via a `setRadix()` or similar method to the Builder.
Because it's a bit of a cross-cut between `Unit` and `Quantity` I would see `MixedRadix` in the `function` package. It also interacts with others like `Calculus` which you also placed there.

**Comment by keilw** [comment]
There's a problem with `MixedQuantityFormat`, that it is typed and therefore a new Format would have to be instantiated for every single quantity:
`MixedRadixQuantityFormat format = MixedRadixQuantityFormat.of(mixedRadix);`
would be neutral, but currently it's
`MixedRadixQuantityFormat<Length> format = MixedRadixQuantityFormat.of(mixedRadix);`

The two are too tightly coupled IMO. Especially when you parse() parsing any given string to just one strictly typed Quantity seems extremely cumbersome.

**Comment by andi-huber** [comment]
Let me recap, what we have so far:

1) Fluent construction of immutable, type-safe `MixedRadix<Quantity>` instances.
2) MixedRadix is a holder of a 'primary' unit, that must be present (checked during construction)
3) A MixedRadix's Units are ordered most significant first (checked during construction)
4) A MixedRadix's Units can be converted one into each other using only 'linear' UnitConverters (checked during construction)
5) Radix arithmetic has 2 challenges: `Composition` and `Decomposition`. Let `r[i]` be the radices and `x[i]` be the coefficients:
- Composition is to calculate a `sum = ... + r[1] r[0] x[2]  + r[0] x[1]  + x[0] `
- Decomposition is to reconstruct all `x[i]` from a given `sum` and given radices `r[i]`, such that only `x[0]` (the least significant coefficient) is fractional, but all other coefficients x[i | i>0] are non-fractional
6) If a mixed-radix quantity has both non-fractional number parts and rational radices the radix arithmetic can be done entirely with BigInteger, otherwise a mix of BigInteger and BigDecimal is used
7) Formatting in mixed-radix notation involves for the coefficients 2 independent NumberFormats
- a (real) NumberFormat for the least significant (trailing) number part (coefficient), which is fractional
- a (integer) NumberFormat for the non-fractional number parts (coefficients)
8) When formatting to mixed-radix notation, the information which Unit is the 'primary' Unit gets lost.
9) It is straight forward how to parse a mixed-radix string, if the parser 'knows' the MixedRadix that was used to format the string.
10) It is yet unclear how to parse a mixed-radix string, without 'knowledge' of the MixedRadix that was used to format the string

**Comment by andi-huber** [comment]
What I want to do next:
- find out how to best handle composition and decomposition of negative numbers; can we do this without surprises for the user?
- test-cases for composition and decomposition of negative numbers
- fill-in missing Java-doc in `MixedRadix`

I'm feeling less competent/motivated to deal with the formatting/parsing stuff.

**Comment by keilw** [comment]
Thanks, the help is greatly appreciated, especially dealing with all the calculations.

What we have to ask ourselves is, whether or not `MixedQuantityFormat` can be a true `QuantityFormat` although it currently implements it? If the need for strong typing by the constructor or of() methods remain and it is not possible to format or parse `Mass` with a `Length` formatter, this would be quite restrictive and could not be retrieved via `FormatService`.

**Comment by keilw** [comment]
I would still find a setPrimary() instead of mixPrimary() less confusing, it would throw
```java
// IllegalStateException: the primary unit is not part of the MixedRadix
MixedRadix<Time> mixedRadix = MixedRadix.ofPrimary(HOUR).mixPrimary(MINUTE);
```

Although it is more of a detail.

Since `MixedRadix` is primarily a replacement for `MixedUnit` or similar, but does not hold a `Quantity` state, what about allowing it to create a `Quantity` and `MixedQuantity` alike? The class currently called `CompoundQuantity`, but if we follow the notion of e.g. Wolfram:
```
In[2]:= | ✕Quantity[MixedMagnitude[{1, 2}], MixedUnit[{"Feet", "Inches"}]]
```
This would be done via:
```java
MixedRadix<Length> mixedRadix = MixedRadix.of(FOOT, INCH);
MixedQuantity<Length> mixedQuantity = mixedRadix.createMixedQuantity(1, 2);
```
The `MixedQuantity` (currently `CompoundQuantity`) has perfect formatting and parsing support already. Effectively it is not much more than an iteration over the elements. Pretty much like the Lambda in `MixedQuantityFormat`, but it relies on a strong type which has no real value for the formatting or parsing as such. And especially when parsing any given string like "6 ft 10 in" there is almost no way to guess the `Quantity` type before you needed to create a strongly typed `QuantityFormat` implementation like `MixedQuantityFormat<Length>`.

**Comment by andi-huber** [comment]
I'm answering in 2 parts

1) We could simply rename `mix*` to `add*`:

Finding a good naming is always opinionated. I feel using `add*` seems to be an improvement.

```java

// primary = defined by convention, either LEADING (=HOUR) or TRAILING (=SECOND)
MixedRadix<Time> mixedRadix = MixedRadix.of(HOUR).add(MINUTE).add(SECOND);

// primary = HOUR
MixedRadix<Time> mixedRadix = MixedRadix.ofPrimary(HOUR).add(MINUTE).add(SECOND);
        
// primary = MINUTE
MixedRadix<Time> mixedRadix = MixedRadix.of(HOUR).addPrimary(MINUTE).add(SECOND);

// primary = SECOND
MixedRadix<Time> mixedRadix = MixedRadix.of(HOUR).add(MINUTE).addPrimary(SECOND);

// IllegalStateException: a primary unit was already defined
MixedRadix<Time> mixedRadix = MixedRadix.ofPrimary(HOUR).addPrimary(MINUTE);
```

2) Regarding `CompoundQuantity`

I thought, the whole point of introducing the `MixedRadix` utility is to make a `CompoundQuantity` obsolete. Maybe I'm missing something?

**Comment by dautelle** [comment]
Hi,
I thought we agreed that MixedUnit/CompoundUnit/MixedRadix were not a good
idea and multi-units should be dealt with at the quantity level (e.g.
Compound Quantity). Furthermore I strongly believe that CompoundQuantity is
useful only if inners quantities are kept separated otherwise as Martin
wrote once, it is a formatting issue and we do not need a new quantity
class for that.
Thank you,
Jean-Marie

**Comment by keilw** [comment]
About 1. I have no strong preferences but the `of()` vs. `ofPrimary()` with an extra `addPrimary()` or `mixPrimary()` seems a bit cumbersome in the API. To use the JSR 310 or `StringBuilder` analogy we could also call it append() because it does not add things except maybe in the sense of a `Collection`. 

About 2. it somehow solves the whole `Unit.mix()` dilemma and replaces a `MixedUnit` or `CompoundUnit`, but it holds no value. It is more a special kind of `QuantityFactory` (was even thinking of adding such a method, see https://github.com/unitsofmeasurement/unit-api/issues/187) 

We made good progress by descoping this topic from the API and regardless of `CompoundQuantity`, `MixedRadix` or a combination of both, I think we do not necessarily need extra API methods yet for them. If there's no consent on how to deal with it even on an implementation level, we may have to descope it from the RI and postpone this. There is too much the API and other features already accomplished to jeopardize that or miss World Metrology Day. Better get the 2.0 release out, since this topic is no longer affecting the API, it can always be added and improved later.

**Comment by keilw** [comment]
@dautelle So what do you suggest we should do here, drop the `MixedRadix` idea?
The biggest issue I have compared to the otherwise fully functional `https://github.com/unitsofmeasurement/indriya/blob/master/src/main/java/tech/units/indriya/quantity/CompoundQuantity.java` is, that the `MixedQuantityFormat` is too tightly coupled to `MixedRadix` and only works for one particular type or `Quantity` at a time. 
"inners quantities are kept separated" can you elaborate on that? `CompoundQuantity` requires no `CompoundUnit` or anything else, it simply builds compatible quantities and holds them together, so they can also be formatted. Call it `QuantityCollection`, I don't really care because that is effectively what it does. The formatting of that `CompoundQuantity` does not manipulate values like `MixedRadix.visitQuantity()` does, and therefore it's not necessary to restrict the quantity type for a particular `format()` or `parse()` operation.

What do you suggest on that (also other @unitsofmeasurement/experts, @unitsofmeasurement/contributors)?

**Comment by dautelle** [comment]
Hello,
Compound quantities are more than a collection of separated quantities, it
implements the quantity interface. Its actual quantity for conversion
purpose is the sum of itd parts stated in SI unit. What is important here
is to keep the original units and integer values. Many applications /
datasets need quantities for exact storage in their own units, not for
conversion! For example: newQuantity(1, DAY_CALENDAR).with(48, HOUR) is
neither 3 days nor 72 hours !! And i don't want to convert it in any.
Thank you,
Jean-Marie

**Comment by dautelle** [comment]
I am fine with the class name CompositeQuantity for compound quantities and
the method add(...) instead of with(...) to construct the composite. What
matters most  is that the v2 is useful and not overly complex. Most users
need strong typing not conversion 😉
Also please do not remove the capability for implementations to have
quantities of compatible dimensions (e.g. Mass and Energy) otherwise
JScience implementation will not be able to support relativistic models.
Thank you,
Jean-Marie

**Comment by andi-huber** [comment]
> Compound quantities are more than a collection of separated quantities, it implements the quantity interface

It does not! I do agree with Werner, its currently (!) just a collection of Quantites. You cannot do quantitiy arithmetic (add, multiply) with `CompositeQuantity`. You need to convert a `CompositeQuantity` to a `Quantity` before you can do any arithmetic on it. When doing this you lose information on all the concrete radix coefficients.

> Many applications / datasets need quantities for exact storage in their own units

`MixedRadix` as a utility allows to decompose a quantity into mixed radix coefficients, without loss of precision. Especially when using integer coefficients and integer radices!

**Comment by andi-huber** [comment]
@dautelle would it make sense to interpret `CompositeQuantity` not as a `Quantity`, but as a vector in the mathematical sense. Namely a vector of Quantities without any constraints on the Unit type for the individual vector elements. This is a totally different topic and does not have anything to do with 'mixed-radix' notation.

We then could provide some basic vector arithmetic such as, but nothing more:

```java
class CompositeQuantity ... { // no generic type constraint
...

 // component-wise quantity addition (method 1)
 // allowed only if both vectors have component-wise compatible Units
 CompositeQuantity add(CompositeQuantity that);

 // component-wise quantity addition (method 2)
 // allowed only if (same as above)
 CompositeQuantity add(Quantity ... components);

 // component-wise scalar multiplication with single factor
 CompositeQuantity multiply(Number factor);

}
```

Also the formatting/parsing of such a vector would be straight forward.

**Comment by dautelle** [comment]
This is were i disagree. A composite quantity is the quantity made of the
sum of its parts. Its unit is the SI unit and is value is the sum of its
values in SI unit. There is no issue in implementing the quantity
interface. The only difference with a standard quantity is that its parts
are all accessible separately with no conversion.

**Comment by andi-huber** [comment]
Sure, we would need to rename this proposed 'vector' class to something like `QuantityVector`. But my point stays the same. You cannot do quantity arithmetic on a composite quantity without losing information on the concrete radix coefficients. Not sure, why we have different views on that?

**Comment by desruisseaux** [comment]
Hello @dautelle . There is two aspects I see in this discussion:

1. Whether we need to differentiate "1 day 50 hours" from "3 days 2 hours". What would be the use case for the need to differentiate those 2 representations?
2. The potential for rounding errors when converting e.g. "0.6666… days" into "16 hours". Given that there is different ways of achieving that goal (storing components separately, or using `BigDecimal`, or using ratio of `long` numbers), shouldn't it be considered as an implementation details as long as an interface is able to provide those components, regardless if they were stored or computed?

**Comment by dautelle** [comment]
Hi @Martin, my first concern is to be able to keep original units in a
composite quantity. For example a calendar day is not 24 hours but *24.0000006
hours. Samething with non SI units, e.g miles, foot, inch...*
*Most applications i know never perform any conversion and they work always
with the same subset of units.  *

**Comment by keilw** [comment]
@dautelle, @andi-huber, @desruisseaux Thanks for the lively discussion everyone. I would not completely outrule that `CompoundQuantity` (or `MixedQuantity` whatever you prefer) may extend/implement `Quantity`. The whole conversion and decomposition should not be done by `QuantityFormat`, nor should it be necessary to create a new `QuantityFormat` for every operation if the general pattern (e.g. `"Number Unit, … possible compound elements"`) won't change. If you parse a well-formed homogenous CSV file you're no supposed to create a new parser for every line either.

@desruisseaux I'm not sure, about the need for 1. either, at least it has lower priority IMO. 2. on the other hand is already solved for every Quantity conversion regardless if that quantity was part of a "compound" or exists independently.

**Comment by keilw** [comment]
@andi-huber I'm afraid `VectorQuantity` or `QuantityVector` is something much different, see https://www.youtube.com/watch?v=eODoCJJ7A8U or https://www.youtube.com/watch?v=_QC42w0npwQ. So I would stick to `CompoundQuantity`, (`CompositeQuantity` or `MixedQuantity` would also work but we gotta make up our minds soon)

**Comment by keilw** [comment]
@andi-huber Is there anything more to do here?

**Comment by andi-huber** [comment]
I'm not convinced that we need to provide `CompoundQuantities` with this RI. I think users might be able to easily write their own implementations, if they have such use-cases.

My suggestion is to rather focus on 'critical' parts of this RI. Eg. we have now support for multi-radix notation. At least with this everything seems well defined and we get the arithmetic right.

But I'd not veto keeping `CompoundQuantitiy` if someone want's to do the work.

**Comment by andi-huber** [comment]
... the current implementation of `CompoundQuantity` uses `MixedRadix` to represents its internal state. This is just a workaround, it cannot handle all the use-cases. So there is still work to do, in order to implement correctly.

**Comment by keilw** [comment]
Well how would it otherwise be represented? We discussed many things like an array of Quantity but none really found a consensus, therefore this is a compromise that works to format and parse "1ft, 5in" or similar. The name distinction between `MixedRadix` and `CompoundQuantity` seems a little odd, but e.g. https://reference.wolfram.com/language/ref/MixedUnit.html does not declare a "MixedQuantity" so using `MixedRadix` as a helper to create a different kind of `Quantity` seems fair.

What other use-cases do you feel are missing?

**Comment by andi-huber** [comment]
Sorry I misread the sources. It appears, that only the `toUnit(Unit)` implementation from interface `QuantityConverter` is problematic. Use-cases not covered by this are if the sequence of Units within a `CompoundQuantity` are not in accordance with mixed-radix convention, as commented in the code: it will throw if units are not in decreasing order of significance.

**Comment by keilw** [comment]
Why would that be problematic? to(Unit) should not throw anything because `CompoundQuantity` already ensures that it only contains compatible elements. Or do you mean creating it with a differently sized `Number` array?

**Comment by andi-huber** [comment]
For example currently you cannot create a compound '5 in, 1 ft', Units must be in decreasing order of significance that is '1 ft, 5 in'. Our mixed-radix implementation is very strict on that, but that allows us to guarantee correct arithmetic.

**Comment by keilw** [comment]
Why would "5 in, 1 ft" not work? Especially for the `CompoundUnit.to()` method it should not matter what order the units are in.

**Comment by andi-huber** [comment]
Right it should work for the `CompoundQuantity`, but does not because in its current implementation we use `MixedRadix` to to the calculation, which will fail if you throw "5 in, 1 ft" at it.

**Comment by keilw** [comment]
I see when trying to test "70 cm, 1 m", why do we need that? I mentioned the distinction between "primary" and "the rest" makes little sense and gets it a bit complicated.

**Comment by andi-huber** [comment]
That is because, with our mixed-radix implementation, arithmetic is always done with the least significant unit, as this is the one with the highest precision. By (our) convention, the least significant unit is also always the last one in the list of units that make up a mixed radix instance. I believe there is good reason for mixed-radix notation to strictly enforce this convention.

**Comment by keilw** [comment]
I think at least the `java.time.Period` class has a similar restriction, so `Period.parse("P1Y2M3D")` works while `Period.parse("P2M3D1Y")` fails. Same for `java.time.Duration`, so all implementations of `java.time.temporal.TemporalAmount` in the JDK require a particular order from "most significant" to "least significant", so I think we could live with a similar restriction here, too.

**Comment by keilw** [comment]
It should however be properly documented. 
Maybe the `MixedRadix<Q> mixedRadix = MixedRadix.of(getUnits());` part could already be moved into the construction of a `CompoundQuantity` and throw a `MeasurementException` (or `IllegalArgumentException`, but since this is very UoM specific I think `MeasurementException` is better here) at construction time already instead of just when trying to convert?

**Comment by andi-huber** [comment]
With PR #232, I'm allowing `CompoundQuantity` to be more flexible than `MixedRadix`. Eg. it allows the compound "70 cm, 1 m" or even  "70 cm, 20 cm", why not.

For that to work I've removed method `Quantity<Q> get(Unit<Q> unit)` in `CompoundQuantity`.

**Comment by keilw** [comment]
If that works, great, thanks.

**Comment by andi-huber** [comment]
Guess I left work for you to correct some of the java-doc in `CompoundQuantity` :)

**Comment by keilw** [comment]
I merged what's there so far. Thanks. I did notice, at least via `SimpleQuantityFormat` quantities that contain  a `RationalNumber` get formatted something like "(29 ÷ 24) ft", does this work both ways? Otherwise formatting and parsing may be slightly out of sync here.

**Comment by andi-huber** [comment]
I've not implemented any parsing code, but it would be great to have a parser that interprets "(29 ÷ 24)" as `RationalNumber.of(29, 24)`

**Comment by keilw** [comment]
The problem is, then you'd have to extend or override `NumberFormat` or `DecimalFormat`. This may be a bit beyond the scope of Indriya or the JSR. I wonder, if it could be an alternative to render `toString()` of `RationalNumber` to the `BigDecimal` or double value by default even if it could have a lot of digits and provide something along the lines of the `BigDecimal` method `toEngineeringString()`? Maybe a different name but you get the idea.

**Comment by andi-huber** [comment]
With PR #233 I'm not particularly proud of the new helper class `SimpleNumberScanner` but at least I'm not breaking any existing tests.

**Comment by andi-huber** [comment]
I had prepared a PR for parsing rational number format, but then I realized that this would fail for parsing compounds. 

Giving it a second thought, parsing is much simpler, if we remove white-spaces and parenthesis from the rational number format: eg. `29÷24` or negative `-29÷24`. (Note: Dividend and divisor can always be parsed using `BigInteger`, no specific `NumberFormat` required for that.) 

I'd also suggest to keep the division-sign ÷ a configurable character, we might as well use / as an alternative: eg. `29/24` or negative `-29/24` ... again without any white-spaces.

I'm about to provide a new PR within the next 24 hours or so.

**Comment by keilw** [comment]
Was there a strong reason to remove `Quantity<Q> get(Unit<Q> unit)` from `CompoundQuantity` btw? Anything the call would break or throw? The idea is to get an individual quantity from the compound somewhat similar to `TemporalAmount` in `java.time`.

**Comment by andi-huber** [comment]
> 
> 
> With PR #232, I'm allowing `CompoundQuantity` to be more flexible than `MixedRadix`. Eg. it allows the compound "70 cm, 1 m" or even "70 cm, 20 cm", why not.
> 
> For that to work I've removed method `Quantity<Q> get(Unit<Q> unit)` in `CompoundQuantity`. If thats not desired behavior, I'm happy to revert my last commit.

`MixedRadix` allows Units to occur only once. That's by design. With `CompoundQuantity` I don't know what to do, I've not seen any use-case that would enlighten me.

**Comment by andi-huber** [comment]
To do a little recap on `CompoundQuantity`:
- From my understanding this represents a list of quantities, which all have compatible Units. 
- The list of quantities that goes into the constructor is preserved and not reordered. Hence not a set of Units.
- It allows to sum up all its (list) elements to a single quantity having a desired Unit.
- We currently use it internally for parsing.

Having a look at Java's `TemporalAmount`  and particularly at the implementation `Duration` [1], the API works quite differently, to what we had before with `CompoundQuantity`. The `getUnits()` method returns all supported TemporalUnits, that is those units that are allowed in a call to `long get(TemporalUnit)`, which returns the number value of the temporal amount in given temporal units. Note: This has nothing to do with the Units used to assemble the `TemporalAmount` in first place.

I suggest, unless we can find an agreement on what CompoundQuantity's purpose should be, make it an internal utility class.

[1] https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html

**Comment by keilw** [comment]
That was always the case, in the previous implementation a `TreeMap` was used that keeps the order of creation. `CompoundQuantity` is supposed to do exactly that. And so far a repeating entry was not permitted. A bit higher above you hinted, "70 cm, 10 cm" would be permitted now, if that's the case and it does not break anything, then why not keep it a List, otherwise here we do not want to show all possible units. I also remember JSR 310 is quite limited, but I was not aware of this strange flaw because a TemporalAmount.getSupportedUnits() may well have been a static method.

The JavaDoc "Returns the list of units uniquely defining the value of this TemporalAmount." says something different. If I have a TemporalAmount like "1 y, 2 mon" I would not expect `getUnits()` to contain `MINUTES`. If it did, it would be even more odd than I imagined, but I had reasons to abstain when we voted on it in the EC like `Duration` being mangled into the API.

**Comment by andi-huber** [comment]
Understood. From my perspective nothing breaks, if we keep the latest implementation of `CompoundQuantity` (allowing to have "70 cm, 10 cm") and only slightly update the associated java-doc.

**Comment by andi-huber** [comment]
... but we can also revert to the previous behavior, without breaking anything.

**Comment by keilw** [comment]
Please double-check headers of new files, they just broke the build ;-/ 
It was not obvious at first sight, but seems you cropped the first line.

**Comment by keilw** [comment]
It looks fine now, because the two lists must be in sync, a slightly bulkier but feasible way to access a particular `Quantity` would be `compQuantity.getQuantities().get(compQuantity.getUnits().indexOf(unit);`
See `testLengthsReverse()` in [CompoundQuantityTest.java](https://github.com/unitsofmeasurement/indriya/blob/master/src/test/java/tech/units/indriya/quantity/CompoundQuantityTest.java).
